### DISCUSSION TITLE
TypeError [ERR_INVALID_URL]: Invalid URL

### DISCUSSION THREAD

**Discussion by vinaykulk621** [vercel/next.js#44599] [Help]
Hello, 
Currently in my project when i make a fetch request to my database(mongodb) through 
`/pages/api/getStudent` in api folder i am getting this error
<br>

![image](https://user-images.githubusercontent.com/94120295/210836648-b5557c1a-0e96-4c0f-8896-753036af393c.png)
 
<br>

My `/pages/api` Folder
![image](https://user-images.githubusercontent.com/94120295/210836960-db3c210c-f7d0-4c7c-99e2-6ef654c135a8.png)

<br>

My code in `/pages/api/getStudent` 
```js
import dbConnect from "../../utils/dbConnection";
import Students from '../../models/student'
import jwt from 'jsonwebtoken';

dbConnect()
export default async function (req, res) {
    try {
        const { cookies } = req;
        const token = cookies.ourSiteJwt;
        if (!token) {
            return res.json({ message: "Invalid token!" });
        }
        const { email } = jwt.decode(token)
        const user = await Students.find({ email })
        if (user) {
            const student = user[0]
            return res.json({ student })
        }
    } catch (e) {
        console.log(e);
    }
    return res.json({ "mess": "dail" })
}
```
<br>

PS:
I am using the latest version of next.js(13.1.0)

**Comment by ZeroThe2nd** [comment]
According to #43415, you'll need to provide a full URL.

However this keeps being an issue in my case, as it blocks the pre-rendering process while deploying to Vercel...

**Reply by vinaykulk621** [reply]
I did, no changes!
still getting the error

**Comment by icyJoseph** [comment]
@vinaykulk621 Do you make a call to your API handler inside Server Components or getStaticProps? Or rather, could you share dbConnect's code too?

**Reply by vinaykulk621** [reply]
My `StudentDeatils.jsx` file which is a component which fetches the data from database and populates in on the frontend
```js
"use client";
// imports

export function StudentDetails({ }) {
    // useState() hook used to initialise the card as empty
    // and after fetching data from database update the card field using this hook
    // ex:
    // const [number,setNumber] = useState("")

    // api call
    async function handleStudent() {
        try {
            const res = await fetch("/api/getStudent");
            const ans = await res.json();
            // changing the content of the card on the page
            // setNumber(ans["contact"]);
            // ...
        } catch (e) {
            console.log(e);
        }
    }

    // calling the fetch api as soon as the page renders
    // so the cards on the page is filled with appropriate data
    handleStudent();

    return (
        <div className="flex flex-col space-y-10 text-center">
            <div className="flex flex-row space-x-10">

                {/* card where the data is updated after api fetch call */}
                {/* <Tablets content={number} label="Phone-Number" /> */}
                {/* ...many more cards */}

            </div>
        </div>
    );
}
```

My `dbConnection.js` file that creates a connection to the mongoDB database
```js
import { MongoClient } from 'mongodb'

let uri = process.env.MONGO_URI
let dbName = "test"

let cachedClient = null
let cachedDb = null

if (!uri) {
    throw new Error(
        'Please define the MONGODB_URI environment variable inside .env.local'
    )
}

if (!dbName) {
    throw new Error(
        'Please define the MONGODB_DB environment variable inside .env.local'
    )
}

export default async function connectToDatabase() {
    if (cachedClient && cachedDb) {
        return { client: cachedClient, db: cachedDb }
    }

    const client = await MongoClient.connect(uri, {
        useNewUrlParser: true,
        useUnifiedTopology: true,
    })

    const db = await client.db(dbName)

    cachedClient = client
    cachedDb = db

    console.log("DataBase Hot!!");
    return { client, db }
}

```

**Reply by icyJoseph** [reply]
Alright, so the problem is that the Server Component is pre-rendered, and during pre-rendering there's no server to handle `/api/getStudent` - What's more, in the server, node-fetch or native fetch (IIRC) require absolute URLs. 

So there's two problems:

- Server Components run server side and those require absolute URLS
- There's no API routes available during build step

Though I am a bit confused with your code snippet that has `use client` directive and a server component (async component rather) - I'll assume that's pseudocode, otherwise, there's problems there too, I think.

Anyhow I'd recommend:

```ts
import dbConnect from '../../utils/dbConnection';
import Students from '../../models/student';
import jwt from 'jsonwebtoken';

export async function findStudentByEmail(cookies) {
    dbConnect();
    try {
        const token = cookies.ourSiteJwt;
        if (!token) {
            return {message: 'Invalid token!'};
        }
        const {email} = jwt.decode(token);
        const user = await Students.find({email});
        if (user) {
            const student = user[0];
            return {student};
        }
    } catch (e) {
        console.log(e);
        return {message: 'Error finding student'};
    }
    return {mess: 'dail'};
}
```

And then use this function in Server Components or API routes. To get the cookies object in server components you can use this function https://beta.nextjs.org/docs/api-reference/cookies, which can even get `ourSiteJwt`, so you can directly pass a string to `findStudyByEmail`.

You can also calculate the cookies inside API routes and reuse `findStudyByEmail`.

Calls to `dbConnect` should be safe because it looks like you instantiate a client, and then re-use in successive calls.
